### PR TITLE
Use (parentheses) rather than "quotes" around literal values

### LINKED ISSUE #263: Pick a delimiter for literals other than the double quote
@markusicu wrote in https://github.com/unicode-org/message-format-wg/pull/230#issuecomment-1116903103:

> Speaking of literals, don't enclose them in quotes. One of the stated goals here is to make messages usable as string literals in programming languages. Enclose literals in parentheses `{... {(5) :number} books}` or angle brackets `{... {<5> :number} books}` or pairs of pipe symbols `{... {|5| :number} books}` or similar.

### PR DESCRIPTION
**Pull request by eemeli**
Closes #263 by replacing `"quotes"` with `(parentheses)` as delimiters for literal values. Correspondingly, the String token is renamed as Literal, though it's still expected to always be parsed into the data model as a string value.

Within a Literal, both `(` and `)` require escaping. Strictly this is only required for the closing parenthesis, but not also escaping the start looks odd, and should improve the experience for anyone looking at MF2 source in an editor that uses generic parenthesis-matching highlighting.

### CODE REVIEW

**Review by stasm — APPROVED** [review]

File: spec/syntax.md [hunk 1]
```diff
@@ -300,7 +300,7 @@ This serves 3 purposes:
 - The message should be conveniently embeddable in various programming languages
   without the need to escape characters commonly related to strings, e.g. `"` and `'`.
   Such need may still occur when a singe or double quote is
```
Comment (aphillips) [inline spec/syntax.md:302]: typo `singe` => `single`

**Review by aphillips — APPROVED** [review]
I'm okay with this change, but remain uneasy about using such a common character pair. I buy @stasm's arguments about using separate open/close characters, otherwise I'd suggest using backtick (`U+0060` `). Or maybe use `<` and `>`? Those aren't as common in "normal" text (although they are not HTML friendly)

**Comment by eemeli** [comment]
One challenge with backticks `` ` `` is that they're commonly used to indicate template or "code" strings. For instance in many markdown variants they require doubling the surrounding backticks to be escaped; I needed to use ``` `` ` `` ``` to get one to render right. They're also often visually difficult to distinguish from single quotes, so may make `` `mistakes'`` difficult to spot.

Angle brackets `<` `>` would need to be escaped in XML & HTML environments, which are rather common.

Hence parentheses `(` `)` as something like a least-worst option.

File: spec/syntax.md (hunk 1 above)
Comment (eemeli) [reply spec/syntax.md:302]:
```suggestion
  Such need may still occur when a single or double quote is
```

**Comment by stasm** [comment]
FWIW, in #263 I was trying to say that for literals we might not need separate open/close delimiters, like I'm advocating for in case of patterns. I also have a concern about angle brackets which I shared in #263 (I'm trying to separate the discussion about the design from the discussion about the implementation).

**Comment by mihnita** [comment]
I don't think that round brackets are not much better than `"`

It means that they become "something to escape", because they can be encountered in plain text.
(I don't know if more often than `"` or not)

TLDR: same as Addison, _"remain uneasy about using such a common character pair."_

---

I don't think that `< ... >` works either, as it conflicts with many existing "storage formats"

[Java can use .xml for resources](https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#loadFromXML-java.io.InputStream-), and some prefer them to `.properties` because they are Unicode.
.NET uses .xml files for localization ([resx files](https://docs.microsoft.com/en-us/dotnet/core/extensions/create-resource-files#ResxFiles))
The [Apple stringdict file format](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/StringsdictFileFormat/StringsdictFileFormat.html) used for localization is also xml.
[Android uses xml for translation](https://developer.android.com/guide/topics/resources/string-resource).

Even with HTML-like templating. I know of several localization systems for html that "embed" the localizable strings in html:
```<p i18n=true>Hello {$user}</p>```

**Comment by stasm** [comment]
> It means that they become "something to escape", because they can be encountered in plain text.

There are two "directions" of escaping that we should both be considering.

* Inside: what if I want to use the literal's delimiter as part of the literal?
* Outside: does the literal's delimiter collide with the container's delimiter for strings, or other special characters?

Choosing something else that the single quote, double quote, backtick or angle brackets optimizes for the "outside".

You're right that there may be cases when an "inside" conflict happens. The PR actually makes one of them clear in one of the examples:

```
{(Thu Jan 01 1970 14:37:00 GMT+0100 \(CET\)): datetime weekday=long}
```

Do you think a pipe would be a better choice here? Or any other character?

```
{|Thu Jan 01 1970 14:37:00 GMT+0100 (CET)|: datetime weekday=long}
```

**Comment by stasm** [comment]
FWIW, I think I'm OK with `{(Thu Jan 01 1970 14:37:00 GMT+0100 \(CET\)): datetime weekday=long}` as a trade-off.

**Comment by eemeli** [comment]
Rebased to account for changes in other merged PRs.

@mihnita Did you have concerns about preferring `"..."` over `(...)` for literals?

File: spec/syntax.md [hunk 2]
```diff
@@ -439,26 +439,26 @@ NameChar ::= NameStart | [0-9] | "-" | "." | #xB7
            | [#x0300-#x036F] | [#x203F-#x2040]
 ```
 
-### Quoted Strings
+### Literal
 
-Any Unicode codepoint is allowed in quoted string literals, with the exception of
-`"` (which ends the string literal),
+Any Unicode codepoint is allowed in literals,
```
Comment (markusicu) [inline spec/syntax.md:444]:
```suggestion
Any Unicode code point is allowed in literals,
```

**Review by markusicu — APPROVED** [review]
lgtm

**Review by mihnita — APPROVED** [review]

**Comment by romulocintra** [comment]
@eemeli needs rebase but ok to merge
